core/trino-main/src/main/java/io/trino/sql/planner/RelationPlanner.java

```diff
@@ -1179,9 +1179,14 @@ If casts are redundant (due to column type and common type being equal),
         for (Identifier column : joinColumns) {
             Symbol output = symbolAllocator.newSymbol(column.getValue(), analysis.getType(column));
             outputs.add(output);
-            assignments.put(output, new Coalesce(
-                    leftJoinColumns.get(column).toSymbolReference(),
-                    rightJoinColumns.get(column).toSymbolReference()));
+            if (join.getType() == JoinType.INNER) {
```

Comment:
It would be preferable to implement this as an optimization rule, instead. One benefit is that it would also apply to queries that use coalesce explicitly in a similar manner.